timesketch/lib/analyzers/sigma_tagger.py

```diff
@@ -70,47 +74,128 @@ def run(self):
-                    logging.error(
-                        'Error generating rule in file {0:s}: {1!s}'.format(
-                            rule_file_path, exception))
-                    continue
-
-                for result in results:
-                    logging.info(
-                        '[sigma] Generated query {0:s}'.format(result))
-                    number_of_tagged_events = self.run_sigma_rule(
-                        result, tag_name)
-                    tags_applied[tag_name] += number_of_tagged_events
+
+
+        for dirpath, dirnames, files in os.walk(rules_path):
+
+            if 'deprecated' in dirnames:
+                dirnames.remove('deprecated')
+
+            for rule_filename in files:
+                if rule_filename.lower().endswith('yml'):
+
+                    # if a sub dir is found, append it to be scanned for rules
+                    if os.path.isdir(os.path.join(rules_path, rule_filename)):
+                        logging.error(
+                            'this is a directory, not a file, skipping: {0:s}'.format(
+                                rule_filename))
+                        continue
+
+                    tag_name, _ = rule_filename.rsplit('.')
+                    tags_applied[tag_name] = 0
+                    rule_file_path = os.path.join(dirpath, rule_filename)
+                    rule_file_path = os.path.abspath(rule_file_path)
+                    logging.info('[sigma] Reading rules from {0!s}'.format(
+                        rule_file_path))
+                    with open(rule_file_path, 'r') as rule_file:
+                        try:
+                            rule_file_content = rule_file.read()
+                            parser = sigma_collection.SigmaCollectionParser(
+                            rule_file_content, self.sigma_config, None)
+                            parsed_sigma_rules = parser.generate(sigma_backend)
+                        except NotImplementedError as exception:
+                            logging.error(
+                                'Error generating rule in file {0:s}: {1!s}'.format(
+                                    rule_file_path, exception))
+                            continue
+                        except Exception as exception:
+                            logging.error(
+                                'Error generating rule in file {0:s}: {1!s}'.format(
+                                    rule_file_path, exception))
+                            continue
+
+                        for sigma_rule in parsed_sigma_rules:
+                            try:
+                                simple_counter += 1
+                                # TODO fix that in Sigma hack to get rid of nested stuff
+                                # https://github.com/google/timesketch/issues/1199#issuecomment-639475885
+                                sigma_rule = sigma_rule.replace(".keyword:", ":")
```

Comment:
Why not query .keyword here? .keyword is the raw string in ES, and I guess that the sigma rule depends on that?

Reply:
Ah that is a weird one, because A sigma rule might have nested values, so if you are searching for username.keyword, the field in timesketch would actually be only username. So searching for username.keyword does not match.

The Sigma project also not sure how to solve that: https://github.com/Neo23x0/sigma/issues/705

And there is no way to make nested mappings either in the yaml config file to say:
username.keyword from Sigma matches username in timesketch.

That is why I introduced the (not so nice and elegant) hack here.

Reply:
(https://github.com/Neo23x0/sigma/tree/master/tools#elasticsearch-or-elk)